File: TEO/Mav.UI/Mav.Ui/src/app/services/i-answer.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../models/generic.result.model';
import { AnswerViewModel } from '../models/answer.view.model';
import { AddAnswerViewModel } from '../models/admin/add-answer.view.model';
import { UpdateAnswerViewModel } from '../models/admin/update-answer.view.model';

export interface IAnswerService {

    getAnswerById<T>(id: number): Observable<T>;

    getAnswerList(questionId: number): Observable<UpdateAnswerViewModel[]>;

    createAnswer(newItem: AddAnswerViewModel): Observable<GenericResult<number>>;

    updateAnswer(item: UpdateAnswerViewModel): Observable<GenericResult<number>>;

    insertUpdateAnswer(items: UpdateAnswerViewModel[]): Observable<GenericResult<number>>;

    deleteAnswer(id: number): Observable<GenericResult<number>>;

}

File: TEO/Mav.Services/Abstract/IAnswerService.cs
using Mav.Common.Models;
using Mav.Models.ViewModels;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IAnswerService
    {
        Task<T> GetAnswerById<T>(int id);

        Task<IEnumerable<T>> GetAnswerList<T>(int questionId);

        Task<GenericResult<T>> CreateAnswer<T>(AddAnswerViewModel m);

        Task<GenericResult<T>> UpdateAnswer<T>(UpdateAnswerViewModel m);

        Task<GenericResult<T>> InsertUpdateAnswer<T>(List<UpdateAnswerViewModel> models);

        Task<GenericResult<T>> DeleteAnswer<T>(int id);

        Task<AddUserAssessmentSectionAnswerViewModel> MarkAnswers(AddUserAssessmentSectionAnswerViewModel vm);
    }
}
